src/main/java/org/wildfly/security/auth/server/SubjectUtil.java

```diff
@@ -0,0 +1,167 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wildfly.security.auth.server;
+
+import java.security.Principal;
+import java.security.acl.Group;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.HashSet;
+import java.util.Set;
+
+import javax.security.auth.Subject;
+
+import org.wildfly.common.Assert;
+import org.wildfly.security.auth.principal.NamePrincipal;
+import org.wildfly.security.authz.Attributes;
+import org.wildfly.security.authz.AuthorizationIdentity;
+import org.wildfly.security.authz.MapAttributes;
+import org.wildfly.security.credential.PublicKeyCredential;
+
+/**
+ * Utilities for dealing with {@link Subject}.
+ *
+ * @author <a href="mailto:sguilhen@redhat.com">Stefan Guilhen</a>
+ */
+public final class SubjectUtil {
+
+    /**
+     * Converts the supplied {@link SecurityIdentity} into a {@link Subject}.
+     *
+     * @param securityIdentity the {@link SecurityIdentity} to be converted.
+     * @return the constructed {@link Subject} instance.
+     */
+    public static Subject fromSecurityIdentity(final SecurityIdentity securityIdentity) {
+        Assert.checkNotNullParam("securityIdentity", securityIdentity);
+        Subject subject = new Subject();
+        subject.getPrincipals().add(securityIdentity.getPrincipal());
+
+        // add the 'Roles' group to the subject containing the identity's mapped roles.
+        Group rolesGroup = new SimpleGroup("Roles");
+        for (String role : securityIdentity.getRoles()) {
+            rolesGroup.addMember(new NamePrincipal(role));
+        }
+        subject.getPrincipals().add(rolesGroup);
+
+        // TODO: should we add groups for the other attributes contained in the authorization identity?
+
+        // add a 'CallerPrincipal' group containing the identity's principal.
+        Group callerPrincipalGroup = new SimpleGroup("CallerPrincipal");
+        callerPrincipalGroup.addMember(securityIdentity.getPrincipal());
+        subject.getPrincipals().add(callerPrincipalGroup);
+
+        // check if we have a public key credential and add it to the Subject's public credentials set.
+        IdentityCredentials publicCredentials = securityIdentity.getPublicCredentials();
+        PublicKeyCredential publicKeyCredential = publicCredentials.getCredential(PublicKeyCredential.class);
+        if (publicKeyCredential != null) {
+            subject.getPublicCredentials().add(publicKeyCredential.getPublicKey());
+        }
+
+        // TODO: should we process other public and private credentials?
```

Comment:
Another thought here would be that we could possibly stick the SI on to the Subject *as* a private credential, allowing this conversion to simply pull it out of there if it's already present.